rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,253 @@
+- Connecting pipelines together - The ability to take input from another pipeline.
+- Pipeline quotas - The ability to limit how much data a pipeline can send to a sink.
+- Pre/post-processing of data - The ability to prepare and normalize data.
+- How pipelines should be synchronized between Vector instances.
+
+## Pain
+
+- Vector does not provide the ability to enforce any kind of organizational structure on configuration files, making large configurations painful.
+  - Because of this lack of structure, this is no clear path to achieving delegation and/or isolation of configuration subsections.
+- There is no means for grouping individual components together for observability purposes.
+
+## Proposal
+
+### User Experience
+
+This change will introduce the concept of pipelines to users. A pipeline is defined as:
+
+1. A collection of transforms defined together, outside of the top-level configuration file
+1. Able to draw input from and send output to components defined in the top-level configuration, but are isolated from other pipelines
+1. Having each contained component's internal metrics tagged with the `id` of the pipeline
+
+Pipelines will be loaded from a `pipelines` sub-directory relative to the Vector configuration directory (e.g., `/etc/vector/pipelines`). Therefore, if a user changes the location of the Vector configuration directory they will also change the `pipelines` directory path. They are coupled.
+
+The `pipelines` directory will contain all pipelines represented as individual files. For simplicity, and to ensure users do not overcomplicate pipeline management, sub-directories/nesting are not allowed. This is inspired by Terraform's single-level directory nesting, which has been a net positive for simple management of large Terraform projects.
+
+Each pipeline file is a processing subset of a larger Vector configuration file. Therefore, it follows the same syntax as Vector's configuration (`toml`, `yaml`, and `json`). Each pipeline will have an `id` derived the name of the file without the extension. For example, the pipeline defined in `load-balancer.yml` will have `load-balancer` as its `id`.
+
+Pipelines have access to any components defined in the root configuration directory. For example, if the transform `foo` is defined in `/etc/vector/bar.toml`, it will be accessible by the pipeline `/etc/vector/pipelines/pipeline.toml`, but if a transform `bar` is defined in `/etc/vector/pipelines/another-pipeline.toml`, it will _not_ be accessible by other pipelines.
+
+If no pipeline is defined, Vector behaves as if the feature didn't exist. This way, a configuration from a version without the `pipeline` feature will keep working. If a pipeline file is left empty, Vector behaves as if it doesn't exist.
+
+If any of the following constraints are violated, Vector will error on boot:
```

Comment:
This appears to be the final issue before we can approve this RFC. After talking with a few design partners, a single invalid pipeline should not prevent Vector from starting or reloading. Stopping the world due to one bad pipeline would be very bad for some users. Alternatively, missing data while a pipeline is invalid is also bad (ex: a `kafka` iterator advancing because other pipelines are valid). My preference is to default to allowing Vector to start and logging an error, but provide users an option to change this.

Curious what you think @lukesteensen.

Reply:
If it is the same design partner conversation I'm thinking of, they also wanted Vector to fallback to the previous, working, definition of the failing pipeline. They also seemed to be satisfied with being able to run `vector validate` before merges that would change the config to avoid errors.

I'm inclined to punt on that for now and just keep Vector's current behavior of all-or-nothing for this initial iteration while we learn more.

Reply:
Works for me. Let's punt. @jdrouet can you add this to the "Future work" section?

Reply:
I'd be hesitant to let Vector start up and run while ignoring a portion of the config. In many cases I don't think people will be watching the log output to see that error, so it could effectively be pretty silent.

It's a little further off, but I'd rather catch invalid pipelines in a validation phase before application of a new config, and that workflow could be built into a future fleet management / remote config service. It's just simpler in a lot of ways (e.g. reloading) if we can always assume we're running a full, valid config.

Reply:
I added this to the `Future Improvements` section. I guess I can close this.